AP_FOLLOW: Reset offsets every time Follow mode is entered.

Hwurzburg opened the pull request:
Probably all wrong...no way to SITL test it....no FOLLOW or SIMPLE mode in SITL yet for Rover...
but this is my attempt to allow vehicles switching to FOLLOW mode, to reset the offsets to the parameter value....including current vehicle positions offset...

currently the library uses the param variable itself to store the calculated offsets so, once computed and stored in the param vector, they never can be changed until reboot, or a new offset is loaded via mavlink....

it compiles fine and I think it does the job by using something other than the param var as storage,leaving it intact, but again, no way to test....let me know what I need to change....

PS will add to copter follow mode once its okay and tested for Rover...

peterbarker commented:
On Tue, 1 Oct 2019, Henry Wurzburg wrote:

> Probably all wrong...no way to SITL test it....no FOLLOW or SIMPLE mode in SITL yet for Rover...

Why not?

What happens when you do "mode follow" in SITL for Rover?

Or, perhaps, "mode 6"?

To give it a location of the vehicle it is supposed to be following, I use 
something like this:

alias add m message GLOBAL_POSITION_INT 0 -353633519 1491652464 586000 2000 0 0 0 0

.... and then i use "m" to send the message.  Repeatedly as there's a 
timeout :-)

You could also model an autotest for Rover based on the Copter one:

https://github.com/ardupilot/ardupilot/blob/master/Tools/autotest/arducopter.py#L3398

Whatever you put in such a test is functionality that won't break without 
us knowing about it.

rmackay9 commented:
@Hwurzburg, an alternative implementation would be to have a boolean called something like "reset_offsets_on_exit" and this boolean is set to true if the offsets are all zero when we enter follow mode.  Then when we exit follow mode we check the boolean and reset the offsets back to zero if necessary.  I'm pretty sure we have an exit for modes although follow may not have it right now.  The only problem with my suggestion is if the user modified the offsets by setting the parameters the parameters would return to the value the user input after the flight controller is rebooted.  This is probably a problem with any solution we come up with though

Hwurzburg commented:
am testing on water today....after thinking a bit, I actually like the idea of having to exit the mode in order to activate new offsets if the param is changed or re-initialize vehicle positionally set offsets (offsets param =0)..its consistent behavior between the two cases
..also I think that if a user changes a param, it should be there the next reboot...

Hwurzburg commented:
works perfectly in boat tests, NED and Rel modes....any mavlink param change occurs on next mode entry....except for type offset,which occurs instantly upon change as before....btw I think relative mode for boats is useless, as is....changing direction 180deg is guaranteed to ground the boat....

rmackay9 commented:
@Hwurzburg,

After some further thought, I think there are two related issues with this PR:

- a similar change needs to be done for Copter (I think this PR might break Follow for Copter)
- it's a bit of a "layering violation" for the vehicle code to directly set AP_Follow's internal members like it does when it gets and sets the offsets.  If possible we should keep the logic within AP_Follow itself.  

I've done an alternative implementation here: https://github.com/rmackay9/rmackay9-ardupilot/commits/rover-ap-follow-reset

In this alternative implementation we record whether the offsets were initialised to the distance from the leader (which can only happen if they were zero) and if "yes" then we reset the offsets back to zero when exiting follow mode.  I've also made it work for Copter.

Functionally the result should be very similar except user updates to the offset parameters will take effect immediately.  We still shouldn't recommend this though because users will then see the offsets return to zero when they exit follow mode but if they reboot the flight controller the offsets will become those values that they entered. This is an existing issue though that isn't made any worse by this change.

I haven't actually tested this in SITL yet.

Hwurzburg commented:
yeah, wont work (I actually confirmed on boat) for Rover since the exit method is never called....prolly will for Copter since it calls it...
why would a warm reboot or hard reboot write the re-zero'd param variable to EEPROM? when the user changes it, isnt it stored in EEPROM and variable, and re-read to the variable on reboot?
I must not know how the EEPROM storage works...thought it would be read into a RAM variable...and writes from user would get stored in EEPROM and the RAM variable....must have it wrong...

rmackay9 commented:
@Hwurzburg, the difference is that if a user updates the parameters from the ground station then those changes are saved to the flight controller's eeprom.  In the code though, if we set the parameters value, it is not saved to eeprom unless we use the .set_and_save(xxx) method.

I'll have a look at why the exit isn't working 'cuz that sounds like a bug.

Hwurzburg commented:
changed to your approach, thanks!....added a GCS message when offsets are initialized....helps to tell if the GCS is not sending  heading and/or GPS positions...lets you know if your really entered Follow mode or not and gives an indication that  offsets may have changed when re-entering the mode...

Boat tested...

rmackay9 commented:
Great, merged, thanks!
